### PR TITLE
GitHub Actions: check editorconfig in PRs

### PR DESCRIPTION
**Pull request by zowoq**
Initial discussion in https://github.com/NixOS/nixpkgs/pull/86376#issuecomment-628688974

~~Currently this is a completely untested draft.~~

### CODE REVIEW

**Comment by grahamc** [comment]
I don't think the build is a good idea. We used to use TravisCI to do some builds, but almost every PR had failed bulids because it was too slow and nixpkgs builds were too big. Maybe instead ofborg could build the manual by default?

**Comment by zowoq** [comment]
> Maybe instead ofborg could build the manual by default?

👍 Whatever works best! I'll drop it from this PR.

**Comment by grahamc** [comment]
For the editorconfig check, is there a way for users to do that validation without pushing to github? I guess I'm wondering how we'll know if we're going to merge this and have everything be broken.

**Comment by zowoq** [comment]
> is there a way for users to do that validation without pushing to github? 

Pre-commit hook maybe? But then it's more of a suggestion rather than a requirement?

> I guess I'm wondering how we'll know if we're going to merge this and have everything be broken.

This (should) only check files changed in PRs against master branch so it shouldn't break anything?

**Comment by grahamc** [comment]
I'm not sure, over time it would require every file be updated to match. I'm not saying this is a bad thing or undesirable, but it is maybe a bigger change than you might be expecting :). For example, the first patch to all-packages.nix would probably be a big one!

**Comment by zowoq** [comment]
Yeah, it will probably take a bit of work to get the repo cleaned up enough that we can merge this without breaking every PR but I think most of the changes can be made automatically and submitted in small batches for easier review?

all-packages.nix doesn't seem to need any changes to comply with the current [.editorconfig](https://github.com/NixOS/nixpkgs/blob/master/.editorconfig).

**Comment by zowoq** [comment]
`indent_size` for `.nix` files seems to be the the biggest problem by far:
https://github.com/NixOS/nixpkgs/blob/f6bfb371cba2b5a02f200c2747c1fe2c72bd782f/.editorconfig#L17-L19

I'd suggest disabling it for `nix` files as `nixpkgs-fmt` might be the better tool for this as lots of these files have somewhat odd patterns.

Using actions to enforce everything for non-nix files and indent_style/newlines/whitespace in nix files seems like decent win?

**Comment by Mic92** [comment]
> `indent_size` for `.nix` files seems to be the the biggest problem by far:
> 
> https://github.com/NixOS/nixpkgs/blob/f6bfb371cba2b5a02f200c2747c1fe2c72bd782f/.editorconfig#L17-L19
> 
> I'd suggest disabling it for `nix` files as `nixpkgs-fmt` might be the better tool for this as lots of these files have somewhat odd patterns.

It is probably the most useful knob in this file as it sets up the tabwidth correctly (instead of 4 spaces). Is there a way to make to ignore it in CI?

**Comment by Mic92** [comment]
Also cc @zimbatm and @Rizary 
https://github.com/nix-community/nixpkgs-fmt/issues/56
who want to integrate nixpkgs-fmt into nixpkgs.

**Comment by zowoq** [comment]
> Is there a way to make to ignore it in CI?

Doesn't look like this is possible with any of the actions available via github marketplace.

We could make our own linter using [editorconfig-checker](https://github.com/editorconfig-checker/editorconfig-checker) or similar but I don't want to start on that unless there is a consensus that we'll actually going to use it.

**Comment by Mic92** [comment]
> > Is there a way to make to ignore it in CI?
> 
> Doesn't look like this is possible with any of the actions available via github marketplace.
> 
> We could make our own linter using [editorconfig-checker](https://github.com/editorconfig-checker/editorconfig-checker) or similar but I don't want to start on that unless there is a consensus that we'll actually going to use it.

How about just patching .editorconfig file before in a step the github ci check runs?

**Comment by zowoq** [comment]
> How about just patching .editorconfig file before in a step the github ci check runs?

😆 Yeah, that should work.

**Review by Mic92 — APPROVED** [review]

**Comment by zowoq** [comment]
Hmmm, might be a good idea to do the same thing for docbook as well.

Even if we "fix" all of them first, using this to enforce a style without ofborg checking that it still builds seems like it might cause problems.

**Comment by Mic92** [comment]
@GrahamcOfBorg if I re-call the problem with travis-ci was that we tried to build every package changed by a pr, including mass-rebuilds. Building only the manual seems like a constant-size amount of work.

**Comment by zowoq** [comment]
I guess the editorconfig and the manual build (if we add it here) would need to also check PRs made against staging (and staging-next maybe?) so that those changes don't end up being a mess that needs to be dealt with by whoever tries to merge staging-next back to master. 

Keeping the config in sync across branches might be annoying as well, at least initially until it stabilises.

I suppose this something that `nixpkgs-fmt`  would need to deal with as well.

**Comment by jtojnar** [comment]
DocBook is actually indented with a single space (default value for `xmlformat`).

**Comment by Mic92** [comment]
Let's leave manual builds for a different PR.

**Comment by zimbatm** [comment]
> Also cc @zimbatm and @Rizary who want to integrate nixpkgs-fmt into nixpkgs.

IMO both are complementary. As long as the indent for nix is set to 2 spaces both tools shouldn't conflict. So :+1: for this initiative.

**Comment by zowoq** [comment]
I've split the docbook indent size out into https://github.com/NixOS/nixpkgs/pull/87919 so it can be merged without waiting for this.

File: .github/workflows/editorconfig.yml
Comment (Mic92) [inline .github/workflows/editorconfig.yml:14]:
```suggestion
    - name: disable nix file indent size check
```

**Comment by zimbatm** [comment]
One thing that popped into my mind is that if we start using GitHub Actions, we should carefully consider the [limits that are in place](https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits).

Probably the most relevant one is that there is a total concurrent job per account. The free plan is at 20 (and 5 for macOS). It's possible that the NixOS org has been upgraded already (/cc @grahamc).

**Comment by zowoq** [comment]
> One thing that popped into my mind is that if we start using GitHub Actions, we should carefully consider the [limits that are in place](https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits).

I'm not sure what I should do with this until the limits are addressed as there isn't much point in cleaning up the repo to comply with the `.editorconfig` unless it is going to be enforced.

Moving the check to `ofborg` would be an option to sidestep this but I'm not a rustacean so it's unlikely that I'll be implementing it anytime soon.

**Comment by Mic92** [comment]
20 concurrent jobs sounds fine to me. This check is quite fast. I don't think there would be a large backlog.

**Comment by zowoq** [comment]
> 20 concurrent jobs sounds fine to me. This check is quite fast. I don't think there would be a large backlog. We may have around 500 PRs a week which is should be well below the 20 jobs at the time limit.

Cool.

Having given this a bit more thought today, a concern I have is that we might end up having to disable it because people ~~don't like~~ _hate it_. What I would like to do is test this with reduced set of (hopefully non-controversial) rules first to see how well it works in this repo and how well it is tolerated.

After dropping all of the `indent_size` rules, this set currently has about 2350 errors across a similar number of files. ~2000 are `trim_trailing_whitespace` and the others are `end_of_line` / `insert_final_newline`.

I think as these are easy to fix and that they would also be the most common type of error to be caught by this check, they should make a reasonable test of how viable this is.

If this doesn't cause too many waves I can start on the clean up and we can enable the other rules.

```ini
# EditorConfig configuration for nixpkgs
# http://EditorConfig.org

# Top-most EditorConfig file
root = true

# Unix-style newlines with a newline ending every file, utf-8 charset
[*]
end_of_line = lf
insert_final_newline = true
trim_trailing_whitespace = true
charset = utf-8

# Ignore diffs/patches
[*.{diff,patch}]
end_of_line = ignore
insert_final_newline = ignore
trim_trailing_whitespace = ignore

# see https://nixos.org/nixpkgs/manual/#chap-conventions

# Match nix/perl/python/ruby/shell/docbook files, set indent to spaces
[*.{nix,pl,py,rb,sh,xml}]
indent_style = space
```

**Comment by Mic92** [comment]
> > 20 concurrent jobs sounds fine to me. This check is quite fast. I don't think there would be a large backlog. We may have around 500 PRs a week which is should be well below the 20 jobs at the time limit.
> 
> Cool.
> 
> Having given this a bit more thought today, a concern I have is that we might end up having to disable it because people ~don't like~ _hate it_. What I would like to do is test this with reduced set of (hopefully non-controversial) rules first to see how well it works in this repo and how well it is tolerated.
> 
> After dropping all of the `indent_size` rules, this set currently has about 2350 errors across a similar number of files. ~2000 are `trim_trailing_whitespace` and the others are `end_of_line` / `insert_final_newline`.
> 
> I think as these are easy to fix and that they would also be the most common type of error to be caught by this check, they should make a reasonable test of how viable this is.
> 
> If this doesn't cause too many waves I can start on the clean up and we can enable the other rules.
> 
> ```ini
> # EditorConfig configuration for nixpkgs
> # http://EditorConfig.org
> 
> # Top-most EditorConfig file
> root = true
> 
> # Unix-style newlines with a newline ending every file, utf-8 charset
> [*]
> end_of_line = lf
> insert_final_newline = true
> trim_trailing_whitespace = true
> charset = utf-8
> 
> # Ignore diffs/patches
> [*.{diff,patch}]
> end_of_line = ignore
> insert_final_newline = ignore
> trim_trailing_whitespace = ignore
> 
> # see https://nixos.org/nixpkgs/manual/#chap-conventions
> 
> # Match nix/perl/python/ruby/shell/docbook files, set indent to spaces
> [*.{nix,pl,py,rb,sh,xml}]
> indent_style = space
> ```

Is there a way to differentiate between warnings and hart failures? This way people might not introduce new format inconsistencies but the CI does not fail for existing code.

**Comment by zowoq** [comment]
Unfortunately `.editorconfig` and its linters doesn't seem to offer that much control over how it's enforced or have the concept of new/existing changes.

A couple of linters do have some options e.g. exclude some files or disable a type of check but that's about it.

**Comment by emilazy** [comment]
I think we should at least enforce 2-space indentation for Nix; that's a pretty universal convention across nixpkgs, though I guess there could be Nix syntax corner cases depending on how exactly the linter checks it.

**Comment by zowoq** [comment]
```ini
[*.nix]
indent_style = space
indent_size = 2
```

Just this by itself has ~240000 errors in ~1800 files.

**Comment by dasJ** [comment]
Also what about the tests? The linter forces 4 spaces for some reason. Or are the contents of multiline strings not checked at all?

**Comment by ajs124** [comment]
> Also what about the tests? The linter forces 4 spaces for some reason. Or are the contents of multiline strings not checked at all?

The test linter most likely forces 4 spaces, because it's python and follows [PEP8](https://www.python.org/dev/peps/pep-0008/#indentation).

**Comment by zowoq** [comment]
Pushed a new version as I wasn't really happy with the original.

It using `editorconfig-checker` with an image pushed from their repo, it has an option to disable the indentation check, it's faster and it looks a lot better than original linter.

**Comment by zowoq** [comment]
`-disable-indent-size` is an option in the newest version of `editorconfig-checker` but that adds another ~1500 errors which might be a bit much for a test.

**Comment by Mic92** [comment]
Ok. Let's gives this a shot.

**Comment by zowoq** [comment]
`haskell-modules/hackage-packages.nix` has 230755 indentation errors + 66 whitespace errors.

The rest of the repo only has 19455 errors total including indentation errors.

So I guess it either needs to be excluded from this check or fixed in `hackage2nix` somehow.

File: .github/workflows/editorconfig.yml
Comment (Mic92) [inline .github/workflows/editorconfig.yml:18]: By the way, is this one faster?

```
VERSION="2.0.4"
OS="linux"
ARCH="amd64"
curl -O -L -C - https://github.com/editorconfig-checker/editorconfig-checker/releases/download/$VERSION/ec-$OS-$ARCH.tar.gz && \
tar xzf ec-$OS-$ARCH.tar.gz && \
./bin/ec-$OS-$ARCH -disable-indentation ${{ env.GIT_DIFF }}
```

I have not checked what the container actually adds as a benefit.

Comment (zowoq) [reply .github/workflows/editorconfig.yml:18]: Yeah, it probably is faster. I picked the container so it doesn't need to be bumped to pick up new versions.

Comment (Mic92) [reply .github/workflows/editorconfig.yml:18]: I actually see it as a feature if it does not update itself. It could be easy enough to add though:


```
VERSION=$(curl https://api.github.com/repos/editorconfig-checker/editorconfig-checker/releases/latest | jq -r '.name')
curl -O -L -C - https://github.com/editorconfig-checker/editorconfig-checker/releases/download/$VERSION/ec-linux-amd64.tar.gz && \
tar xzf ec-linux-amd64.tar.gz && \
./bin/ec-linux-amd64 -disable-indentation ${{ env.GIT_DIFF }}
```

Comment (zowoq) [reply .github/workflows/editorconfig.yml:18]: > I actually see it as a feature if it does not update itself.

Yeah good point, sticking to a tag would be safer considering how busy this repo is.

I'll send a PR.

**Comment by marcus7070** [comment]
Could someone with access consider adding a checkbox in the PR template about editorconfig? The first I heard about it was when my PR failed (and GitHub failed to display the error log), which wasn't the nicest first impression.

**Comment by Mic92** [comment]
Do you really want more checkboxes? What was not clear about the error message given by the check? Maybe we should improve on that.

**Comment by marcus7070** [comment]
The error message did not load. Just a spinning little GitHub symbol even though the test was finished and failed. I could access both the earlier setup logs and the later postrun actions logs, but not the actual "editorconfig check" logs. Sorry I can't be more specific, I've fixed and force pushed my PR and can't get back to the error message. I can see editorconfig error messages on other PRs, so it was something specific to mine. 

Incase it helps, the error message should have been something along the lines of
```
pkgs/development/python-modules/cadquery/default.nix:
	25: Trailing whitespace
	38: Trailing whitespace
	71: Trailing whitespace
	71: Wrong amount of left-padding spaces(want multiple of 2)
	78: Trailing whitespace
	78: Wrong amount of left-padding spaces(want multiple of 2)
	82: Trailing whitespace
	82: Wrong amount of left-padding spaces(want multiple of 2)
	95: Trailing whitespace
	95: Wrong amount of left-padding spaces(want multiple of 2)
	97: Trailing whitespace
	97: Wrong amount of left-padding spaces(want multiple of 2)

12 errors found
```

**Comment by Mic92** [comment]
I also justed it in https://github.com/NixOS/nixpkgs/pull/88594

This is the error message:

```
default.nix:
	27: Trailing whitespace

1 errors found
##[error]Process completed with exit code 1.
```

We maybe could rename it from `check editorconfig` to `check codingstyle based on editorconfig` to make it a bit more clearer but apart from that the error message looks understandable to me.

**Comment by marcus7070** [comment]
Just found an email with a link to the logs: https://github.com/NixOS/nixpkgs/runs/698397328
They work fine now.

> add a command to locally reproduce this error 

Yeah, that's the most important part to me.

**Comment by LnL7** [comment]
This spawned a bunch of discussion in #nixos-dev yesterday.

The problem here is that now github actions and oforg both have their own unrelated checks which means if one or the other doesn't run everything still looks good (which happened today). I'm not aware of any particular problems with GitHub actions but after giving it some thought I'm not in favour of depending on multiple CI systems for status checks.

**Comment by grahamc** [comment]
I would like to revert this PR and close #88488. Not because I don't like github actions or feel like my pet project ofborg is threatened, but because I think these checks are actually harming the clearness of the checks. I'll explain:

The editorconfig check puts a big green check next to a commit, and hides the list of checks which passed. If ofborg is down or behind, it is likely people will think ofborg has approved the PR before it did. This could cause, well, all the problems ofborg is there to prevent.

The concrete problem here is this green check appears very early and is misleading, and there is no poka yoke to help people do the right thing.

One fear of #88488 is that it will introduce red X's if it takes a long time to build the manual, or something in the target branch is failing to build. This is reminiscent of the Travis times for me: Travis gave out so few green checkmarks, that if Travis gave a PR a green check-mark, it inspired more "okay what is broken that let travis pass it?" than "okay good to merge".
 
The goal of ofborg, and ever since ofborg became a thing, the "rule" has been green checkmark -> good to merge and red X -> don't merge. 

The editorconfig check errodes the clearness of the green checkmark signal.

The big-rebuild failures we'll certainly have with #88488 errodes the clearness of the red X signal.

The clearness of this signal has been my guiding light on the design of ofborg for years now, and I am very afraid of threatening it. 

A few suggestions that have been brought up to improve this:

1. **make ofborg faster** I'd love to do this, but it doesn't solve this problem if ofborg is seriously delayed or down.
2. **require the ofborg evaluation check for merges** this is really the only solution, but not something we can do today, because of some workflow issues we need to resolve.

That makes me think that, today, we should be undoing this check and closing #88488.

**Comment by zimbatm** [comment]
A related fix is to enable this option in the branch protections: 

image

One of the benefits of doing that is that it lets the user know in advance, before ofborg or GitHub Actions has even executed, which checks are needed to pass before they can merge the PR.

**Comment by grahamc** [comment]
That is not an option right now, @zimbatm. Let's talk about this later? I have information about this, but not a schedule compatible with spending much time on it right now.

**Comment by grahamc** [comment]
note I included that in my message:

> require the ofborg evaluation check for merges this is really the only solution, but not something we can do today, because of some workflow issues we need to resolve.

**Comment by Mic92** [comment]
Maybe we can integrate this check into ofborg itself.

**Comment by zimbatm** [comment]
Got reverted in https://github.com/NixOS/nixpkgs/pull/88608

Sorry @zowoq, I know you put effort into this. Do you mind opening a new PR to re-introduce it? I still think it's a good idea to use it.

**Comment by zowoq** [comment]
Seems I picked a bad time to step away from a computer for a couple of hours.

> Do you mind opening a new PR to re-introduce it?

>> That is not an option right now ... I have information about this, but not a schedule compatible with spending much time on it right now.

I think I'll wait to hear more from @grahamc first before doing anything.

Thanks @Mic92 and everyone else for you help on this.

**Comment by grahamc** [comment]
> Sorry @zowoq <https://github.com/zowoq>, I know you put effort into this. Do you mind opening a new PR to re-introduce it? I still think it's a good idea to use it.

I think it is a good idea, too.

**Comment by zowoq** [comment]
I'd like to look at adding checks for formatting/linting in non-nix files as well, currently we have ~20000 loc in ~500 bash/sh files, ~6500 loc in ~40 python files and ~5000 loc in ~30 perl files.

@grahamc Do you know when you'll be able to provide some detail on the workflow issues that are blocking us from using the required status checks with ofborg?

**Comment by Mic92** [comment]
The alternative would be to have ofborg to build derivations described in nixpkgs itself so that we can add features to the CI without having to modify ofborg itself.

**Comment by zowoq** [comment]
> The alternative would be to have ofborg to build derivations described in nixpkgs itself so that we can add features to the CI without having to modify ofborg itself.

TBH, I don't think adding more things (making us more dependant) on ofborg is a good idea.

**Comment by zowoq** [comment]
> Do you know when you'll be able to provide some detail on the workflow issues that are blocking us from using the required status checks with ofborg?

@grahamc ping

**Comment by Mic92** [comment]
Could there be a dummy github action that like `Waiting for ofborg...`, which ofborg could stop?

**Comment by zowoq** [comment]
PRs can fail on all 3 platforms (grey neutral builds) or the builds might not even start (and not appear as a status) and they would still be green overall.

I don't think using the required status UX would be any worse than what we have now. I'd certainly like to try using them before attempting to implement anything more complicated.

**Comment by Mic92** [comment]
However at least we should not break evaluation. Failed evaluation will mark the build as read.

**Comment by Mic92** [comment]
This github action might be a suitable workaround: https://github.com/NixOS/nixpkgs/pull/93581

**Pull request merged**
